src/domain/task/usecase/register-task/registerTask.ts

```diff
@@ -23,5 +25,5 @@ export const registerTask =
       )
       return
     }
-    dispatch(RegisterTaskActions.taskRegistered(task))
+    dispatch(RegisterTaskActions.taskRegistered(TaskMapper.mapCreateTaskToEntity(task)))
```

Comment:
🤔 As far as I understand, the entity task leaks in the event. It should not.

Reply:
Yes, I know.
But, there is no option here, we need to call a redux action which calls the reducer that updates the state.
For that, we are forced to give the entity as the payload. If not, this will be the reducer itself to build the entity and I don't think this is a good way to achieve that.
The issue here is that the event bus is coupled with redux as a middleware.
So, if we want to avoid the entity leak in the event bus, we need to find a way to 'deserialize' the payload before been published by the bus.
I think it is a deep change that should be addressed in a specific PR, what do you think?

Reply:
@fredericvilcot I don't understand the point. As far as I see, the dispatch + reducer is compatible with ES requirements. You can dispatch an event (which is not an entity) and in the reducer, given what is provided in the event recombine the aggregate. I don't see why you can't.

Reply:
We never handled any business logic in reducers in the previous domains, so payloads are the entities themselves...
To modify that, we need a bigger refactor.

Reply:
Well it's not business logic. The business logic remains in the UC. The reducer just applies (in a pure fashion) the events in the store. So it's really not a big deal, and I think it's a refactor to do here, scoped on this domain (because it's rather simple). It's important to do it right because otherwise you won't be able to scale (and sadly we will continue to break the fundamental rules of the Hexagonal and the Event Sourcing).

Reply:
@ccamel 
Ok, so, if I understand, the UC is simply here a flat pass that receives command payload as input and only dispatches a redux action with the same payload? 
It will be the reducer to gets this payload and build this entity?

Reply:
@fredericvilcot It's a little more subtle. The UC is not a pass through: it is responsible of the logic for validating the command, ensure it is acceptable regarding the current state and finally generates the *different* events. Here's is the business logic: translate a command into several events. And events should be DTOs only (and immutable of course).

The reason why the entity should not be created in the UC is that the different events published can be used to build different models (different Read models). The entity here is only used for the Write model.

Then, in the reducer, we're on the write side, you just apply the events to have the store updated. There's nothing wrong in handling the entities here. It's what a reducer is supposed to do.

Does it make more sense? It's quite difficult to explain in a few words.

Reply:
@ccamel 
I agree with that approach but it remains one interrogation: what about handling possible error when building the entity in the reducer? What about if the entity must load more external context to ne build? 
The reducer is a pure function so we can't apply here side effects like throwing or calling methods that should fail.
So, how can we handle this part of the write side?

Reply:
@fredericvilcot You can't have errors in the reducer. If the event has been published by the UC, then it can't be denied in any case, it must be applied and produce a change (a transition) in the state.

What do you mean by "load more external context"? In the reducer, you can't access other resources than the store (it's a pure function). If you need more to build the entity, it has to be carried by the event and resolved by the UC which can access all the secondary adapters it wants.

Reply:
@ccamel 
Ok, thanks for your explanations.

To respect the ES principles, I will move invariants from builders to the UC to ensure that the emitted event is foolproof.

So, as a result, the reducer will only have to worry about aggregate's recomposition by managing the state.

Therefore, IMHO, entity builders can no longer be used directly in the UC except for testing because the invariants must be checked before the event is propagated and therefore before the entity itself is built.

If you agree with that, I can resolve this comment and implement the changes.